compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java

```diff
@@ -906,11 +912,169 @@ private boolean checkSimilarMatchPatterns(BLangMatchPattern firstPattern, BLangM
+            if (!checkSimilarConstMatchPattern(firstSimpleMatchPattern.constPattern,
+                    secondSimpleMatchPattern.constPattern)) {
+                return false;
+            }
+        } else if (!(firstSimpleMatchPattern.constPattern == null && secondSimpleMatchPattern.constPattern == null)) {
+            return false;
+        }
+
+        if (firstSimpleMatchPattern.wildCardMatchPattern != null
+                && secondSimpleMatchPattern.wildCardMatchPattern == null) {
+            return false;
+        }
+        if (firstSimpleMatchPattern.wildCardMatchPattern == null
+                && secondSimpleMatchPattern.wildCardMatchPattern != null) {
+            return false;
+        }
+
+        if (firstSimpleMatchPattern.varVariableName != null
+                && secondSimpleMatchPattern.varVariableName == null) {
+            return false;
+        }
+        if (firstSimpleMatchPattern.varVariableName == null
+                && secondSimpleMatchPattern.varVariableName != null) {
+            return false;
+        }
+
+        return true;
+    }
+
+    private boolean checkSimilarErrorCauseMatchPattern(BLangErrorCauseMatchPattern firstErrorCauseMatchPattern,
+                                                       BLangErrorCauseMatchPattern secondErrorCauseMatchPattern) {
+        if ((firstErrorCauseMatchPattern != null && secondErrorCauseMatchPattern == null)
+                || (firstErrorCauseMatchPattern == null && secondErrorCauseMatchPattern != null)) {
+            return false;
+        }
+        if (firstErrorCauseMatchPattern == null) {
+            return true;
+        }
+        if (!checkSimilarSimpleMatchPattern(firstErrorCauseMatchPattern.simpleMatchPattern,
+                secondErrorCauseMatchPattern.simpleMatchPattern)) {
+            return false;
+        }
+        return checkSimilarErrorMatchPattern(firstErrorCauseMatchPattern.errorMatchPattern,
+                secondErrorCauseMatchPattern.errorMatchPattern);
+    }
+
+    private boolean checkSimilarErrorFieldMatchPatterns(BLangErrorFieldMatchPatterns firstErrorFieldMatchPatterns,
+                                                        BLangErrorFieldMatchPatterns secondErrorFieldMatchPatterns) {
+        if ((firstErrorFieldMatchPatterns != null && secondErrorFieldMatchPatterns == null)
+                || (firstErrorFieldMatchPatterns == null && secondErrorFieldMatchPatterns != null)) {
+            return false;
+        }
+        if (firstErrorFieldMatchPatterns == null) {
+            return true;
+        }
+        if (firstErrorFieldMatchPatterns.restMatchPattern != null
+                && secondErrorFieldMatchPatterns.restMatchPattern == null) {
+            return false;
+        }
+        if (firstErrorFieldMatchPatterns.restMatchPattern == null
+                && secondErrorFieldMatchPatterns.restMatchPattern != null) {
+            return false;
+        }
```

Comment:
Aren't the following two patterns similar patterns? The second is unreachable, right?

```ballerina
public function main() {
    error x = error("err", i = 1);

    match x {
        error (var y, i = 1) => {
        }

        error (var y, i = 1, ...var z) => {
        }
    }
}
```